frontend/src/modules/otherlocales/components/Locales.js

```diff
@@ -9,19 +9,56 @@ import Translation from './Translation';
 
 import type { Navigation } from 'core/navigation';
 import type { LocalesState } from '..';
+import type { UserState } from 'core/user';
 
 
 type Props = {|
     otherlocales: LocalesState,
+    user: UserState,
     parameters: Navigation,
     updateEditorTranslation: (string) => void,
+    preferredCount: number,
 |};
 
 
 /**
  * Shows all translations of an entity in locales other than the current one.
  */
 export default class Locales extends React.Component<Props> {
+    /**
+     * Orders the list of locales. The list starts with prefered locales,
+     * in order as they are defined. The remaining locales follow in the
+     * given (alphabetic) order.
+     */
+    sortByPreferred() {
+        const { otherlocales, user } = this.props;
+        const translations = otherlocales.translations;
+
+        if (!user.isAuthenticated) {
+            return translations;
+        }
+
+        const preferredLocales = user.preferredLocales.reverse();
+
+        return translations.sort((a, b) => {
+            let indexA = preferredLocales.indexOf(a.code);
+            let indexB = preferredLocales.indexOf(b.code);
+
+            if (indexA === -1 && indexB === -1) {
+                return a > b;
+            }
+            else if (indexA < indexB) {
+                return 1;
+            }
+            else if (indexA > indexB) {
+                return -1;
+            }
+            else {
+                return 0;
+            }
+        });
+    }
```

Comment:
I think this should also be a selector. There will be a lot of cases where you re-do that computation but the outcome will be exactly the same, because the inputs didn't change.

Reply:
In general, you should avoid having logic in your `render` methods as much as possible. Put as much as makes sense in either actions, reducers or selectors.